Added a cheat code system, and more

ZackWilde27 opened the pull request:
After calling SDK.Init(), call SDK.CheatCodes.Hook(), then you can add cheat codes with SDK.CheatCodes.Add()

ZackWilde27 committed:
Added a cheat codes system

- After calling SDK.Init(), call SDK.CheatCodes.Hook(), then you can add cheat codes with SDK.CheatCodes.Add()

Sewer56 reviewed (commented):
Congrats on your first PR; now let's see if you can learn to amend a PR 😉

- `CheatCodeManager` can be a static, as there's no use to having multiple instances of it.
- Add XML documentation for `CheatCodeManager`, since that is public API.
- As a general style guide, move the inner types `NewCheatCode` to the bottom of the file.
- The function definitions i.e. `CheckForCheatCodesPtr`, `CheatCodeCallback` should go to a new file in `FlatOut2.SDK/Functions`
- Cheats are rarely added, but game code is better left untouched if possible. So check if the hook for cheats was applied in the `CheatCodeManager.Add` function, and if it hasn't apply it once and never again.
- You shouldn't need to apply the function attribute to `NewCheckForCheatCodes`.
- For `NewCheckForCheatCodes`, that must be static, else you're introducing a hidden `this` parameter (which would take ecx register). That risks register corruption.
- In `CharPtrToString`, you're making a new string for every character read, which means you're doing a new heap allocation and copy every character. You would normally want to use a `StringBuilder` in C# if doing the reading manually. 
    - But ideally just use `Marshal.PtrToStringAnsi` from the standard library.

Happy gaming!

ZackWilde27 commented:
Should I document the private functions too, or only the public?

ZackWilde27 committed:
Documentation and fixes

Made it all static, and hooking is automatic. Instead of CheatCodes.Add() it's now CheatCodeManager.Add(), if you have the 'using FlatOut2.SDK.API' directive

ZackWilde27 committed:
Final (maybe) pass

I just decided to just put the Hook code directly in Add, and some small tweaks to documentation

ZackWilde27 committed:
Nope, at least 1 more fix

The money gets rounded to 27272728 (I'm guessing due to float imprecision from Lua) I was planning on settling with that but I changed my mind

ZackWilde27 committed:
Ok, last one (hopefully)

Some more changes to documentation and removing the unnecessary 'using' from SDK.cs

ZackWilde27 committed:
While I'm here

- Added the Car struct

- Added onto existing structs, like adding ReadOnlyPosition to Player, and CareerOdometer and StuntID to RaceInfo

- Added a new Helpers class to the API to make common things a lot easier to do. I started with a helper that hooks RenderSky to get a function called every frame

- Added a GetPlayers function to Info.Race, that converts the raw pointer from RaceInfo into a C# array

ZackWilde27 committed:
6th time's the charm

Forgot to make the field offsets in the Car struct all caps, much better

ZackWilde27 committed:
Fixes to the GetPlayers function

- The NumCars on the HostObject appears to be used for multiple things, so it could be wrong sometimes and cause a crash
I found a different place to get the number from

- I also updated the check so it doesn't need to store a pointer anymore

ZackWilde27 committed:
Dang it more fixes

I decided to remove my GetPlayers function, it was still causing rare crashes, I'll try and figure that out some other time

ZackWilde27 committed:
Last minute fixes

- Added a summary for the Helpers class itself
- Realized the CareerOdometer is on the PlayerProfile, not the RaceInfo
- Moved the unsafe on the Car to just the Owner
- Removed the new using statement in Info.cs

Sewer56 commented:
Will check at end of tomorrow :p

ZackWilde27 commented:
Sounds good

ZackWilde27 committed:
Last last minute additions

I found some more stuff

- Added DerbyType and DerbyTypeOverride to the RaceInfo struct, and documented what each mode does

- Added the Garage struct and added it to the PlayerProfile

- Found where the PlayerHost's pointer is stored, so it has an Instance member now

- Added IsMultiplayer, and LastPlayer to the PlayerHost

- Added the BOOL type from win32 so that certain bool-like members can be treated like bools

- Found what the StuntTypes are actually called, so I updated the struct and file name

- Added Info.Controller, for reading the keyboard

- Added Info.Profile, to query the PlayerProfile

ZackWilde27 committed:
Forgot to remove some testing stuff

Also added a bit of documentation and removed the unsafes that weren't needed

ZackWilde27 committed:
More small changes

- Renamed LastPlayer to NextPlayer, makes more sense that way

- Changed NitroRegen on RaceInfo from an int to the BOOL type

- Some of the BOOL types had documentation from when they were ints, which isn't really needed anymore

ZackWilde27 committed:
Don't mind me, more fixes

I changed the parity array from bools to bytes, I don't know if the conversion to bool affects speed at all, but just in case

ZackWilde27 committed:
Added back CarClass, and documentation changes

ZackWilde27 committed:
Tweaked the cheat code system

- Used CharSet.Unicode to skip Marshalling the code

- The hooked function now returns a BOOL so I can return true instead of 1

- Capitalized the documentation on AnyKeysDown() and AnyKeysPressed()

ZackWilde27 committed:
Re-added GetPlayers

It uses a span to be more efficient than the last version

I couldn't have a span of pointers, so the easiest way to cast it is to give the iterator the Player* type, like:
foreach (Player* player in Info.Race.GetPlayers())

Also some small documentation changes and additions

ZackWilde27 committed:
Small change

Realized the CareerOdometer is on the Garage, not the PlayerProfile

Also removed some using statements that weren't needed

ZackWilde27 committed:
Alright last change I think

Signed my comment about the implicit operators, should probably do that since it's not my repo

ZackWilde27 committed:
Fixes and additions

- Added GameRulesOverride to the RaceInfo struct, and documented what each set of rules do

- Fixed a bug with HookPerFrame() where multiple mods calling it would end up crashing the game

- Added WorldMatrix to the Car struct, and since the position is just the last row of the matrix I changed the description and made it a Vector3

ZackWilde27 committed:
Changed velocity to Vector3s

Makes more sense to have it that way, makes doing match with them much easier

Also added IsWrecked to the cars, to make checking that easier
